File: go_backend/controllers/chat_controller.go
package controllers

import (
	"net/http"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/realtime-chat/go_backend/config"
	"github.com/realtime-chat/go_backend/models"
	"go.mongodb.org/mongo-driver/bson/primitive"
)

// GetChatToken generates a token for chat services using Stream Chat API
// @Summary Get Stream Chat token
// @Description Generate a token for authenticating with Stream Chat API
// @Tags Chat
// @Accept json
// @Produce json
// @Security BearerAuth
// @Success 200 {object} models.ChatToken "Successfully generated token"
// @Failure 401 {object} map[string]string "Unauthorized - User not authenticated"
// @Failure 500 {object} map[string]string "Internal server error"
// @Router /api/chat/token [get]
func GetChatToken(c *gin.Context) {
	// Get the user ID from the context (set by middleware)
	userIDObj, exists := c.Get("userID")
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "User ID not found in context"})
		return
	}

	userID, ok := userIDObj.(primitive.ObjectID)
	if !ok {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to parse user ID"})
		return
	}

	// Convert ObjectID to string for Stream Chat
	userIDStr := userID.Hex()

	// Generate Stream Chat token
	token, err := config.StreamClient.CreateToken(userIDStr, time.Now().Add(24*time.Hour))
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to generate Stream token", "details": err.Error()})
		return
	}

	c.JSON(http.StatusOK, models.ChatToken{Token: token})
}

// SaveMessage saves a chat message to the database
// @Summary Save a chat message
// @Description Save a new chat message to the database
// @Tags Chat
// @Accept json
// @Produce json
// @Security BearerAuth
// @Param message body models.MessageRequest true "Message data"
// @Success 201 {object} models.Message "Successfully saved message"
// @Failure 400 {object} map[string]string "Bad request - Invalid message data"
// @Failure 401 {object} map[string]string "Unauthorized - User not authenticated"
// @Failure 500 {object} map[string]string "Internal server error"
// @Router /api/chat/messages [post]
func SaveMessage(c *gin.Context) {
	// Extract message data from request
	var message models.Message
	if err := c.ShouldBindJSON(&message); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	// Get user ID from context (set by middleware)
	userIDObj, exists := c.Get("userID")
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "User ID not found in context"})
		return
	}

	userID, ok := userIDObj.(primitive.ObjectID)
	if !ok {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to parse user ID"})
		return
	}

	// Set message properties
	message.UserID = userID
	message.CreatedAt = time.Now()
	message.ID = primitive.NewObjectID()

	// Here you would save the message to MongoDB
	// For now, we'll just return a success response
	c.JSON(http.StatusCreated, message)
}

File: backend/src/controllers/chat.controller.js
import { generateStreamToken } from "../lib/stream.js";

/**
 * @swagger
 * /chat/token:
 *   get:
 *     summary: Get Stream Chat token for video calls
 *     tags: [Chat]
 *     security:
 *       - cookieAuth: []
 *     responses:
 *       200:
 *         description: Stream token generated successfully
 *         content:
 *           application/json:
 *             schema:
 *               type: object
 *               properties:
 *                 token:
 *                   type: string
 *                   example: "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9..."
 *       401:
 *         description: Unauthorized
 *       500:
 *         description: Internal server error
 */
// This function is used for video calls, while Go backend handles text chat
export async function getStreamToken(req, res) {
  try {
    const token = generateStreamToken(req.user.id);

    res.status(200).json({ token });
  } catch (error) {
    console.log("Error in getStreamToken controller:", error.message);
    res.status(500).json({ message: "Internal Server Error" });
  }
}
